### PR TITLE
Enabling 2nd Backend git 

### PR DESCRIPTION
**Pull request by fbrbovic**
This pull request allows for having confd poll two backends at the same time and provide the data from both in the same templates.

To use it : 
set `-backend2=` to backend of your choice
example:
```
confd -backend=etcd -backend2=vault -auth-type2=token -auth-token2=${AUTH_TOKE} -node2=http://10.22.1.102:8200
```
set any additional flags related to the backend2  (flags which end with '2' example `-node2=` )

set backend2 keys in the toml file, by setting `keys2=`
example:
```
[template]
src = "test.tmpl"
dest = "/tmp/test.conf"
mode = "777"
keys = [
  "/message",
]
keys2 = [
  "/secret/api/external/docker_hub",
]
```

use the backend 2 vars in the template file by prefixing the keys with `backend2:`
example: 
```
{{getv "backend2:/secret/api/external/docker_hub/username"}}
```

This request was stemmed from issue : https://github.com/kelseyhightower/confd/issues/414

### CODE REVIEW

**Comment by michaellopez** [comment]
I think this looks promising

**Comment by odedlaz** [comment]
The idea is awesome, but I think we need to find a generic way of doing this instead of allowing just one more backend. plus, there's a lot of duplication in the code which will make changing and bugs fixes difficult in the future.

**Comment by fbrbovic** [comment]
@odedlaz Yes I agree there is definitely a lot that can be done to generalize adding of additional backends. I needed a simple solution to enable a 2nd backend right away for my use case.
I've considered working on a more generalize way but it would require a lot more work since we would also need to ensure that it would also be backwards compatible with existing flags. 

One option I considered was to keep the same flags. Example: ```--backend=```
but enable some kind of namespacing/aliasing capability inside the flag values like : ```--backend="backend1:etcd,backend2:vault,backend3:rancher"```, but it would have taken me a lot longer to implement, and I didn't have time. 

Question is also beyond 2nd or possible third backend, how many uses cases would need more then 2 or 3 simultaneous backends to justify added complexity?

**Comment by bacongobbler** [comment]
A couple of problems I see with this approach are key collision, maintainability and the added complexity of this PR.

For key collision, keys are now namespaced by their backend (`getv "backend1:/foo"`) in this PR. Not only is this backwards-incompatible, but this also means that the templates must have intimate knowledge of the way confd was configured to fetch their keys. The key paths are meant to be a filesystem-like key/value store similar to etcd's key/value layout. Even if we reverted this to make all keys added to the root path (`/`), we would then have the issue of collision detection in two backends that have the same path (e.g. etcd has `/foo` and vault has `/foo`). This alone is a huge issue that would need to be properly resolved.

To address maintainability, if we allow N backends to be configured, what's wrong with configuring N+1? If we're already going down this approach, a user should be able to configure any number of backends. 

From #414 it just looks like the problem's more about how to retrieve values from vault securely while also storing non-critical information in another backend, not retrieving values from more than one backend at a time. I don't see how it'd be useful to read keys from both etcd and stackengine, for example.

To summarize I don't think this is the right way to solve #414. It adds a significant amount of added complexity to solve one use case, which doesn't seem like it's worth the hassle to maintain IMO.

**Comment by fbrbovic** [comment]
@bacongobbler  As far as I understand templates always need to have intimate knowledge of how confd is configured because confd allows to set a prefix and that changes the way vars are accessed. If you change the prefix then templates will now not work just like if you were to swap the backends, as you described it. 

Either way backend swapping problem can be solved if we were to implement using the name of the backend as the namespace for the variable instead of the word "backend". 
So we would have ```getv "vault:/foo"```  instead of ```getv "backend1:/foo"```.   This way order the backends doesn't matter. Of course we would not be able to use two same backend types together. 

Also I don't see how using a namespace  (name:/path/path) causes any problems or incompatibilities. 
Paths still stay paths, we are only adding optional namespace prefixes to them, nobody is converting paths into windows style backslashes (\)

But lets say that namespaces were big showstopper we could always create a new method for accessing multi-backend systems like ```(getbv "vault" "/foo/morefoo")```  or ```(getbv "backend1" "/foo/morefoo")```

Anyway there seems to be many options to solve these problems, as long as confd community cares about having multiple backends capability. With all that said I can understand the argument of it being more hassle then it's worth and confd not wanting to go in such direction.

For my use case it was either write a simple custom system that behaves like confd or extend confd with multi backend capability. Obviously I chose the latter and figured someone else might also be in need of similar functionality.

**Comment by odedlaz** [comment]
@fbrbovic, as far as I know @kelseyhightower doesn't care about backward compatibility until we reach the 1.0 milestone.

**Comment by bacongobbler** [comment]
#414 is not something we're interested in supporting, so I'm closing this.
